File: packages/sdk-python/src/clawtoken/dao.py
"""DAO Governance API — proposals, voting, delegation, treasury, timelock."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from clawtoken.http import AsyncHttpClient, HttpClient


class DaoApi:
    """Synchronous DAO Governance API."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    # ── Proposals ──────────────────────────────────────────────────────

    def list_proposals(self, *, status: str | None = None) -> dict[str, Any]:
        """List proposals, optionally filtered by status."""
        params: dict[str, Any] | None = {"status": status} if status else None
        return self._http.get("/api/dao/proposals", params)

    def get_proposal(self, proposal_id: str) -> dict[str, Any]:
        """Get a single proposal by ID."""
        return self._http.get(f"/api/dao/proposals/{quote(proposal_id, safe='')}")

    def create_proposal(self, **kwargs: Any) -> dict[str, Any]:
        """Create a new proposal."""
        return self._http.post("/api/dao/proposals", kwargs)

    def advance_proposal(self, proposal_id: str, **kwargs: Any) -> dict[str, Any]:
        """Advance a proposal to a new status."""
        return self._http.post(
            f"/api/dao/proposals/{quote(proposal_id, safe='')}/advance", kwargs
        )

    # ── Voting ─────────────────────────────────────────────────────────

    def get_votes(self, proposal_id: str) -> dict[str, Any]:
        """Get votes for a proposal."""
        return self._http.get(
            f"/api/dao/proposals/{quote(proposal_id, safe='')}/votes"
        )

    def vote(self, **kwargs: Any) -> dict[str, Any]:
        """Cast a vote on a proposal."""
        return self._http.post("/api/dao/vote", kwargs)

    # ── Delegation ─────────────────────────────────────────────────────

    def delegate(self, **kwargs: Any) -> dict[str, Any]:
        """Set delegation to another DID."""
        return self._http.post("/api/dao/delegate", kwargs)

    def revoke_delegation(self, **kwargs: Any) -> dict[str, Any]:
        """Revoke a delegation."""
        return self._http.post("/api/dao/delegate/revoke", kwargs)

    def get_delegations(self, did: str) -> dict[str, Any]:
        """Get delegations for a DID."""
        return self._http.get(f"/api/dao/delegations/{quote(did, safe='')}")

    # ── Treasury ───────────────────────────────────────────────────────

    def get_treasury(self) -> dict[str, Any]:
        """Get current treasury status."""
        return self._http.get("/api/dao/treasury")

    def deposit(self, **kwargs: Any) -> dict[str, Any]:
        """Deposit into the treasury."""
        return self._http.post("/api/dao/treasury/deposit", kwargs)

    # ── Timelock ───────────────────────────────────────────────────────

    def list_timelock(self) -> dict[str, Any]:
        """List timelock entries."""
        return self._http.get("/api/dao/timelock")

    def execute_timelock(self, action_id: str, **kwargs: Any) -> dict[str, Any]:
        """Execute a timelocked action."""
        return self._http.post(
            f"/api/dao/timelock/{quote(action_id, safe='')}/execute", kwargs
        )

    def cancel_timelock(self, action_id: str, **kwargs: Any) -> dict[str, Any]:
        """Cancel a timelocked action."""
        return self._http.post(
            f"/api/dao/timelock/{quote(action_id, safe='')}/cancel", kwargs
        )

    # ── Params ─────────────────────────────────────────────────────────

    def get_params(self) -> dict[str, Any]:
        """Get governance parameters and thresholds."""
        return self._http.get("/api/dao/params")


class AsyncDaoApi:
    """Asynchronous DAO Governance API."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    # ── Proposals ──────────────────────────────────────────────────────

    async def list_proposals(self, *, status: str | None = None) -> dict[str, Any]:
        """List proposals, optionally filtered by status."""
        params: dict[str, Any] | None = {"status": status} if status else None
        return await self._http.get("/api/dao/proposals", params)

    async def get_proposal(self, proposal_id: str) -> dict[str, Any]:
        """Get a single proposal by ID."""
        return await self._http.get(
            f"/api/dao/proposals/{quote(proposal_id, safe='')}"
        )

    async def create_proposal(self, **kwargs: Any) -> dict[str, Any]:
        """Create a new proposal."""
        return await self._http.post("/api/dao/proposals", kwargs)

    async def advance_proposal(self, proposal_id: str, **kwargs: Any) -> dict[str, Any]:
        """Advance a proposal to a new status."""
        return await self._http.post(
            f"/api/dao/proposals/{quote(proposal_id, safe='')}/advance", kwargs
        )

    # ── Voting ─────────────────────────────────────────────────────────

    async def get_votes(self, proposal_id: str) -> dict[str, Any]:
        """Get votes for a proposal."""
        return await self._http.get(
            f"/api/dao/proposals/{quote(proposal_id, safe='')}/votes"
        )

    async def vote(self, **kwargs: Any) -> dict[str, Any]:
        """Cast a vote on a proposal."""
        return await self._http.post("/api/dao/vote", kwargs)

    # ── Delegation ─────────────────────────────────────────────────────

    async def delegate(self, **kwargs: Any) -> dict[str, Any]:
        """Set delegation to another DID."""
        return await self._http.post("/api/dao/delegate", kwargs)

    async def revoke_delegation(self, **kwargs: Any) -> dict[str, Any]:
        """Revoke a delegation."""
        return await self._http.post("/api/dao/delegate/revoke", kwargs)

    async def get_delegations(self, did: str) -> dict[str, Any]:
        """Get delegations for a DID."""
        return await self._http.get(f"/api/dao/delegations/{quote(did, safe='')}")

    # ── Treasury ───────────────────────────────────────────────────────

    async def get_treasury(self) -> dict[str, Any]:
        """Get current treasury status."""
        return await self._http.get("/api/dao/treasury")

    async def deposit(self, **kwargs: Any) -> dict[str, Any]:
        """Deposit into the treasury."""
        return await self._http.post("/api/dao/treasury/deposit", kwargs)

    # ── Timelock ───────────────────────────────────────────────────────

    async def list_timelock(self) -> dict[str, Any]:
        """List timelock entries."""
        return await self._http.get("/api/dao/timelock")

    async def execute_timelock(self, action_id: str, **kwargs: Any) -> dict[str, Any]:
        """Execute a timelocked action."""
        return await self._http.post(
            f"/api/dao/timelock/{quote(action_id, safe='')}/execute", kwargs
        )

    async def cancel_timelock(self, action_id: str, **kwargs: Any) -> dict[str, Any]:
        """Cancel a timelocked action."""
        return await self._http.post(
            f"/api/dao/timelock/{quote(action_id, safe='')}/cancel", kwargs
        )

    # ── Params ─────────────────────────────────────────────────────────

    async def get_params(self) -> dict[str, Any]:
        """Get governance parameters and thresholds."""
        return await self._http.get("/api/dao/params")

File: packages/sdk/src/dao.ts
/**
 * DAO Governance API — proposals, voting, delegation, treasury, timelock.
 */
import type { HttpClient, RequestOptions } from './http.js';
import type {
  DaoProposalListResponse,
  DaoProposalResponse,
  DaoCreateProposalParams,
  DaoAdvanceProposalParams,
  DaoVotesResponse,
  DaoVoteCastParams,
  DaoDelegateSetParams,
  DaoDelegateRevokeParams,
  DaoDelegationsResponse,
  DaoTreasuryResponse,
  DaoTreasuryDepositParams,
  DaoTimelockListResponse,
  DaoTimelockActionParams,
  DaoTimelockCancelParams,
  DaoParamsResponse,
  DaoTxResult,
  DaoProposalStatus,
} from './types.js';

export class DaoApi {
  constructor(private readonly http: HttpClient) {}

  // ── Proposals ──────────────────────────────────────────────────────

  /** List proposals, optionally filtered by status. */
  async listProposals(
    status?: DaoProposalStatus,
    opts?: RequestOptions,
  ): Promise<DaoProposalListResponse> {
    const params = status ? { status } : undefined;
    return this.http.get<DaoProposalListResponse>(
      '/api/dao/proposals',
      params as Record<string, string> | undefined,
      opts,
    );
  }

  /** Get a single proposal by ID. */
  async getProposal(
    proposalId: string,
    opts?: RequestOptions,
  ): Promise<DaoProposalResponse> {
    return this.http.get<DaoProposalResponse>(
      `/api/dao/proposals/${encodeURIComponent(proposalId)}`,
      undefined,
      opts,
    );
  }

  /** Create a new proposal. */
  async createProposal(
    params: DaoCreateProposalParams,
    opts?: RequestOptions,
  ): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>('/api/dao/proposals', params, opts);
  }

  /** Advance a proposal to a new status. */
  async advanceProposal(
    proposalId: string,
    params: DaoAdvanceProposalParams,
    opts?: RequestOptions,
  ): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>(
      `/api/dao/proposals/${encodeURIComponent(proposalId)}/advance`,
      params,
      opts,
    );
  }

  // ── Voting ─────────────────────────────────────────────────────────

  /** Get votes for a proposal. */
  async getVotes(
    proposalId: string,
    opts?: RequestOptions,
  ): Promise<DaoVotesResponse> {
    return this.http.get<DaoVotesResponse>(
      `/api/dao/proposals/${encodeURIComponent(proposalId)}/votes`,
      undefined,
      opts,
    );
  }

  /** Cast a vote on a proposal. */
  async vote(params: DaoVoteCastParams, opts?: RequestOptions): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>('/api/dao/vote', params, opts);
  }

  // ── Delegation ─────────────────────────────────────────────────────

  /** Set delegation to another DID. */
  async delegate(params: DaoDelegateSetParams, opts?: RequestOptions): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>('/api/dao/delegate', params, opts);
  }

  /** Revoke a delegation. */
  async revokeDelegation(
    params: DaoDelegateRevokeParams,
    opts?: RequestOptions,
  ): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>('/api/dao/delegate/revoke', params, opts);
  }

  /** Get delegations for a DID. */
  async getDelegations(
    did: string,
    opts?: RequestOptions,
  ): Promise<DaoDelegationsResponse> {
    return this.http.get<DaoDelegationsResponse>(
      `/api/dao/delegations/${encodeURIComponent(did)}`,
      undefined,
      opts,
    );
  }

  // ── Treasury ───────────────────────────────────────────────────────

  /** Get current treasury status. */
  async getTreasury(opts?: RequestOptions): Promise<DaoTreasuryResponse> {
    return this.http.get<DaoTreasuryResponse>('/api/dao/treasury', undefined, opts);
  }

  /** Deposit into the treasury. */
  async deposit(
    params: DaoTreasuryDepositParams,
    opts?: RequestOptions,
  ): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>('/api/dao/treasury/deposit', params, opts);
  }

  // ── Timelock ───────────────────────────────────────────────────────

  /** List timelock entries. */
  async listTimelock(opts?: RequestOptions): Promise<DaoTimelockListResponse> {
    return this.http.get<DaoTimelockListResponse>('/api/dao/timelock', undefined, opts);
  }

  /** Execute a timelocked action. */
  async executeTimelock(
    actionId: string,
    params: DaoTimelockActionParams,
    opts?: RequestOptions,
  ): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>(
      `/api/dao/timelock/${encodeURIComponent(actionId)}/execute`,
      params,
      opts,
    );
  }

  /** Cancel a timelocked action. */
  async cancelTimelock(
    actionId: string,
    params: DaoTimelockCancelParams,
    opts?: RequestOptions,
  ): Promise<DaoTxResult> {
    return this.http.post<DaoTxResult>(
      `/api/dao/timelock/${encodeURIComponent(actionId)}/cancel`,
      params,
      opts,
    );
  }

  // ── Params ─────────────────────────────────────────────────────────

  /** Get governance parameters and thresholds. */
  async getParams(opts?: RequestOptions): Promise<DaoParamsResponse> {
    return this.http.get<DaoParamsResponse>('/api/dao/params', undefined, opts);
  }
}
